### DISCUSSION TITLE
Multi-Stage Docker Build : Docker.DotNet.DockerApiException : Docker API responded with status code=Conflict

### DISCUSSION THREAD

**Discussion by kolexinfos** [testcontainers/testcontainers-dotnet#858] [General]
I have a .Net Core 6 Project which I am trying to setup integration tests for using TestContainers.
I am using a Dockerfile generated by Visual Studio and this builds my image and I am able to run the container locally and the web api works as expected but when I run the same DockerFile in a TestContainers project I get the exception :

**_Docker.DotNet.DockerApiException : Docker API responded with status code=Conflict, response={"message":"Container 51d3ead2c1f194aef0c5876cbc29affe89c81c40b142131cd037e6eba8f5e624 is not running"}_**

Here is my Dockerfile
**_FROM [mcr.microsoft.com/dotnet/aspnet:6.0](http://mcr.microsoft.com/dotnet/aspnet:6.0) AS base
WORKDIR /app
EXPOSE 80
EXPOSE 443
FROM [mcr.microsoft.com/dotnet/sdk:6.0](http://mcr.microsoft.com/dotnet/sdk:6.0) AS build
WORKDIR /
COPY ./nuget.config /nuget.config
COPY ["MyProjectDir/MyProject.csproj", "MyProject/"]
COPY ["Common/Common.csproj", "Common/"]
RUN dotnet restore "MyProjectDir/MyProject.csproj"
COPY . .
WORKDIR "/MyProject"
RUN dotnet build "MyProject.csproj" -c Release -o /app/build
FROM build AS publish
RUN dotnet publish "MyProject.csproj" -c Release -o /app/publish /p:UseAppHost=false
FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENTRYPOINT ["dotnet", "MyProject.dll"]_**

One thing I observed though is it seems the TestContainers instance is failing to run the final build stage of the Dockerfile as I only see the Publish Stage image when the process is completed and then get an exception.

Is there anything I am missing or is there an issue with Multi Container Builds?

**Comment by HofmeisterAn** [comment]
AFAIK multi-stage works fine, we use it [here](https://github.com/testcontainers/testcontainers-dotnet/blob/5152cffa097b660292989cddd490f80dcfefa994/examples/WeatherForecast/Dockerfile).

> I am using a Dockerfile generated by Visual Studio [...]

Usually, the build context is not set correct (directory relative to the files required for the Docker build). See the following example: https://github.com/testcontainers/testcontainers-dotnet/issues/610#issuecomment-1299960403. Please make sure you pass the correct paths to the builder. If you still struggle with the configuration, can you please share your `ImageFromDockerfileBuilder()` configuration?

In addition you can set a custom logger (`TestcontainersSettings.Logger`) that forwards all log messages to get more information why your build fails.

**Comment by kolexinfos** [comment]
Here is my configuration

```
await new ImageFromDockerfileBuilder()
                  .WithName(this)
                  .WithDockerfileDirectory(CommonDirectoryPath.GetSolutionDirectory(), string.Empty)
                  .WithDockerfile("MyProject/Dockerfile")
                  .WithBuildArgument("RESOURCE_REAPER_SESSION_ID", ResourceReaper.DefaultSessionId.ToString("D"))
                  .WithDeleteIfExists(false)
                  .Build()
                  .CreateAsync()
                  .ConfigureAwait(false);
```
                  
  I already attached a Debugger, I briefly see a Container created with the Image I tagged in my Docker Desktop in a Running state but one weird thing happening now is the Test in VisualStudio never completes and the Log file keeps growing, I had to stop it. Here is a link to the logs https://pastebin.com/2uWFTtxN

**Reply by HofmeisterAn** [reply]
Based on the logs, it appears that the image is being built successfully. However, the `UntilPortIsAvailable(443)` wait strategy does not seem to complete, see:

    /bin/sh -c true && (cat /proc/net/tcp{,6} | awk '{print $2}' | grep -i :443 || nc -vz -w 1 localhost 443 || /bin/bash -c '</dev/tcp/localhost/443')

This command tests whether the port is available or not. It continues to run until the port becomes available. There could be two reasons for the failure: either your application is not starting or listening on port 443, or the command is incompatible with your Docker image. For example, the executables required to test whether the port is available might be missing in the image.
